wicket-core/src/main/java/org/apache/wicket/pageStore/FilePageStore.java

```diff
@@ -0,0 +1,569 @@
+		try
+		{
+			FileChannel channel = FileChannel.open(file.toPath(), StandardOpenOption.CREATE,
+				StandardOpenOption.TRUNCATE_EXISTING, StandardOpenOption.WRITE);
+			try
+			{
+				ByteBuffer buffer = ByteBuffer.wrap(data);
+				channel.write(buffer);
+			}
+			finally
+			{
+				IOUtils.closeQuietly(channel);
+			}
+		}
+		catch (IOException ex)
+		{
+			log.warn("cannot store page data for session {} page {}", identifier, pageId, ex);
+		}
+
+		setPageType(file, pageType);
+	}
+
+	private void checkMaxSize(String identifier)
+	{
+		File[] files = folders.get(identifier, true).listFiles();
+		Arrays.sort(files, new LastModifiedComparator());
+
+		long total = 0;
+		for (int f = 0; f < files.length; f++)
+		{
+			File candidate = files[f];
+
+			total += candidate.length();
+
+			if (total > maxSizePerSession.bytes())
+			{
+				if (!Files.remove(candidate))
+				{
+					log.warn("cannot remove page data for session {} page {}", identifier,
+						candidate.getName());
+				}
+			}
+		}
+	}
+
+	protected SessionAttribute getSessionAttribute(IPageContext context, boolean create)
+	{
+		SessionAttribute attribute = context.getSessionAttribute(KEY);
+		if (attribute == null && create)
+		{
+			attribute = new SessionAttribute(applicationName, context.getSessionId());
+			context.setSessionAttribute(KEY, attribute);
+		}
+		return attribute;
+	}
+
+	/**
+	 * Attribute held in session.
+	 */
+	static class SessionAttribute implements Serializable, HttpSessionBindingListener
+	{
+
+		private final String applicationName;
+
+		/**
+		 * The identifier of the session, must not be equal to {@link Session#getId()}, e.g. when
+		 * the container changes the id after authorization.
+		 */
+		public final String identifier;
+
+		public SessionAttribute(String applicationName, String sessionIdentifier)
+		{
+			this.applicationName = Args.notNull(applicationName, "applicationName");
+			this.identifier = Args.notNull(sessionIdentifier, "sessionIdentifier");
+		}
+
+
+		@Override
+		public void valueBound(HttpSessionBindingEvent event)
+		{
+		}
+
+		@Override
+		public void valueUnbound(HttpSessionBindingEvent event)
+		{
+			FilePageStore store = FILE_STORES.get(applicationName);
+			if (store == null)
+			{
+				log.warn(
+					"Cannot remove data '{}' because disk store for application '{}' is no longer present.",
+					identifier, applicationName);
+			}
+			else
+			{
+				store.removeFolder(identifier);
+			}
+		}
+	}
+
+	public class LastModifiedComparator implements Comparator<File>
+	{
+
+		@Override
+		public int compare(File f1, File f2)
+		{
+			return (int)(f2.lastModified() - f1.lastModified());
+		}
+
+	}
+
+	@Override
+	public String getContextIdentifier(IPageContext context)
+	{
+		return getSessionAttribute(context, true).identifier;
+	}
+
+	@Override
+	public Set<String> getContextIdentifiers()
+	{
+		Set<String> identifiers = new HashSet<>();
+
+		for (File folder : folders.getAll())
+		{
+			identifiers.add(folder.getName());
```

Comment:
It seems we store all files in one flat folder ?! This caused problems before and a logic to split by prefix has been introduced.

Reply:
+1